scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -142,43 +144,51 @@
-                     nil :task-failed)
-        status {:task-id {:value instance-id}
-                :state task-state
-                :reason (container-status->failure-reason compute-cluster instance-id pod-status job-container-status)}
-        exit-code (some-> job-container-status .getState .getTerminated .getExitCode)]
-    (write-status-to-datomic compute-cluster status)
-    (when exit-code
-      (sandbox/aggregate-exit-code (:exit-code-syncer-state compute-cluster) instance-id exit-code))
-    ; Must never return nil, we want it to return non-nil so that we will retry with writing the state to datomic in case we lose a race.
-    ; Being in the (completed,*) state, will cause us to delete the pod, transitioning to (completed,missing), and thence
-    ; to deleting from the map, into (missing,missing) state.
-    {:cook-expected-state :cook-expected-state/completed}))
+  (when-not (synthetic-pod-job-uuid pod)
+    (let [instance-id (-> pod .getMetadata .getName)
+          pod-status (.getStatus pod)
+          ^V1ContainerStatus job-container-status (get-job-container-status pod-status)
+          ; We leak mesos terminology here ('task') because of backward compatibility.
+          task-state (case (:state synthesized-state)
+                       :pod/failed :task-failed
+                       :pod/succeeded :task-finished
+                       :pod/unknown :task-failed
+                       :pod/waiting :task-failed ; Handle the (:cook-expected-state/running,:pod/waiting) case.
+                       nil :task-failed)
+          status {:task-id {:value instance-id}
+                  :state task-state
+                  :reason (container-status->failure-reason compute-cluster instance-id pod-status job-container-status)}
+          exit-code (some-> job-container-status .getState .getTerminated .getExitCode)]
+      (write-status-to-datomic compute-cluster status)
+      (when exit-code
+        (sandbox/aggregate-exit-code (:exit-code-syncer-state compute-cluster) instance-id exit-code))))
+  ; Must never return nil, we want it to return non-nil so that we will retry with writing the state to datomic in case we lose a race.
+  ; Being in the (completed,*) state, will cause us to delete the pod, transitioning to (completed,missing), and thence
+  ; to deleting from the map, into (missing,missing) state.
+  {:cook-expected-state :cook-expected-state/completed})
 
 (defn handle-pod-started
   "A pod has started. So now we need to update the status in datomic."
   [compute-cluster {:keys [pod]}]
-  (let [instance-id (-> pod .getMetadata .getName)
-        ; We leak mesos terminology here ('task') because of backward compatibility.
-        status {:task-id {:value instance-id}
-                :state :task-running}]
-    (write-status-to-datomic compute-cluster status)
-    {:cook-expected-state :cook-expected-state/running}))
+  (when-not (synthetic-pod-job-uuid pod)
```

Comment:
Do we log anywhere that the synthetic pod started running?

Reply:
Yes, it will get logged by the state machine when the state transition happens